SAMD fixes (PinI2C)

awgrover opened the pull request:
Changes to enable samd. So far, tested with:

    controller: adafruit metro express (m0)
    target: adafruit itsybitsy m0
    examples/Pin_control
    edited firmware/firmware_modules.h to include PinI2C_firmware.h
    dPinOut = 13
    dPinIn = 12
    aPinOut = 3
    aPinIn = 14 // A0

Switch and potentiometer worked.

I intend on testing a M4 soon, and trying to figure out `reset()`

awgrover committed:
use correct type for analogReference()

awgrover committed:
Merge branch 'doc-review' into fixes-1

awgrover committed:
firmware/PinI2C_firmware.h:
allow analogReference() for samd
src/PinI2C.cpp
static cast

awgrover committed:
firmware/firmware.ino:
finish enabling samd support.
use static cast for eAnalogReference.
disable EEPROM unless on known architectures.
arduino ide reformat made some changes.
wait on Serial for "native" usb architectures.

(cherry picked from commit c98233850cd2b9d2dd5afd533cacdb14e5329c49)

awgrover commented on firmware/PinI2C_firmware.h:
```diff
@@ -77,11 +77,11 @@
           if (i == 1) { //1 uint8_t
             uint8_t mode; bufferIn->read(mode);
-            analogReference(mode);
+            analogReference(static_cast<eAnalogReference>(mode));
```
please see change in method signature. The compiler was complaining about the wrong type.

ftjuh replied:
Forget my below comment. Your code does compile for ESPs, but not for Uno/Nano:

```
PinI2C_firmware.h: In function 'void processMessage(uint8_t)':
PinI2C_firmware.h:84:41: error: 'eAnalogReference' does not name a type; did you mean 'analogReference'?
             analogReference(static_cast<eAnalogReference>(mode));
                                         ^~~~~~~~~~~~~~~~
                                         analogReference
exit status 1
'eAnalogReference' does not name a type; did you mean 'analogReference'?

```
Never heard of that enum, but makes sense. Would be nice if the Arduino docs would properly document each functions' parameter types, as it's not mentioned at https://www.arduino.cc/reference/en/language/functions/analog-io/analogreference/ and is hard to google.

ftjuh replied:
And frankly I don't understand why it compiles for ESPs as I can only find eAnalogReference used in the samd core:

```
~/.arduino15/packages$ find . -type f -exec grep -H --color eAnalogReference '{}' \;
./arduino/hardware/samd/1.8.13/cores/arduino/wiring_analog.c:void analogReference(eAnalogReference mode)
./arduino/hardware/samd/1.8.13/cores/arduino/Arduino.h:typedef enum _eAnalogReference
./arduino/hardware/samd/1.8.13/cores/arduino/Arduino.h:} eAnalogReference ;

```

awgrover replied:
Damn, I thought I tested for Uno/avr's. What do you think is the best strategy? I think this caused a compile fail for samd.

awgrover commented on firmware/firmware.ino:
```diff
@@ -19,7 +19,10 @@
+#if defined(ARDUINO_ARCH_AVR) || defined(ARDUINO_ARCH_ESP32) || defined(ARDUINO_ARCH_ESP8266)
+#define USE_EEPROM
 #include <EEPROM.h>
+#endif
```
Disable EEPROM unless on known good architectures. The `#define` means we don't have to repeat the exact `#if` expression elsewhere.

ftjuh replied:
ok

awgrover replied:
There is a flash mechanism that is almost equivalent. The major difference is that a stored flash value is reset whenever new code is uploaded. I'll explore this eventually.

https://www.arduino.cc/reference/en/libraries/flashstorage/

awgrover commented on firmware/firmware.ino:
```diff
@@ -186,14 +193,15 @@ void storeI2C_address(uint8_t newAddress)
-   enabled for each unit seperately, 
+   note: The interrupt pin is global, as there is only one shared by all modules
+   and units. However, modules can implement them so that interrupts can be
+   enabled for each unit seperately,
```
Whitespace cleaned up by Arduino IDE. Several other cases.

awgrover commented on firmware/firmware.ino:
```diff
@@ -202,16 +210,16 @@ uint8_t interruptSource = 0xF;
+   @param reason 4-bit value that can be used to differentiate between different
+   interrupt causing events, e.g. end stop hit vs. target reached. 0xF is
+   reserved to signal "unknown reason"
+*/
```
I think the Arduino IDE did that.

awgrover commented on firmware/firmware.ino:
```diff
@@ -286,6 +299,8 @@ void setup()
 #elif defined(ARDUINO_ARCH_ESP32)
   log("ARDUINO_ARCH_ESP32\n");
+#elif defined(ARDUINO_ARCH_SAMD)
+  log("ARDUINO_ARCH_SAMD\n");
```
I'm being optimistic. Only tested 1 device so far.

ftjuh replied:
Great.

awgrover commented on src/PinI2C.cpp:
```diff
@@ -55,9 +55,10 @@ int PinI2C::analogRead(uint8_t pin){
 
-void PinI2C::analogReference(uint8_t mode){
+void PinI2C::analogReference(eAnalogReference mode){
+  static_assert(sizeof(eAnalogReference) == sizeof(uint8_t),"Need argument to analogReference to be same size as uint8_t");
   wrapper->prepareCommand(pinAnalogReferenceCmd, myNum);
-  wrapper->buf.write(mode);
+  wrapper->buf.write(static_cast<uint8_t>(mode));
```
Note the `static_assert`. I think that enforces the enum fitting in a uint8_t.

awgrover commented on firmware/PinI2C_firmware.h:
```diff
@@ -77,11 +77,11 @@
         }
```
I did not attempt to figure out and fix the "reset" (i.e. reboot) command. yet.

ftjuh replied:
Please don't. What is actually needed is modules cleaning up their internal state back to a clean slate, not a hardware reset. 

My early implementation as hardware reset was out of insecurity concerning how to destroy objects initialized with new() in c++. I've learned OO with a proper language which included garbage collection, and just wanted to push that problem for later. I'll do that soon for the existing modules.

awgrover replied:
I see. I'll skip that.

ftjuh reviewed (changes requested):
The eAnalogReference stuff needs to be looked into before a PR.

awgrover committed:
Controller
src/PinI2C.h/cpp:
revert analogReference() to uint8_t. On controller it's uint8_t
examples/Pin_control/Pin_control.ino
Tested with a samd51 and samd21
reformatted by arduino IDE

awgrover committed:
Target
firmware/PinI2C_firmware.h
Add ARDUINO_ARCH_SAMD ARDUINO_ARCH_MEGAAVR supported for
analogReference(), and Servo.h
For ARDUINO_ARCH_SAMD, analogReference(static_cast<eAnalogReference>(mode));

ftjuh committed:
Merge pull request #1 from awgrover/doc-review

Terminology, README links/tweaks/review

ftjuh committed:
added ATtiny "support" in README and Pin_Control.ino example

ftjuh committed:
Merge branch 'feature/attinysupport' into main

ftjuh committed:
minor edits in README, trying to concilate github frontpage with doxygen

ftjuh committed:
minor edits in README

ftjuh committed:
README/doxygen conciliation

ftjuh committed:
README/doxygen conciliation

ftjuh committed:
added TOC to README

ftjuh committed:
removed TOC from README, github doesn't support it (doh)

ftjuh committed:
minor README edit

ftjuh committed:
TM1638lite module added

ftjuh committed:
Merge branch 'TM1638' into main

ftjuh committed:
readme revised, command code ranges corrected

awgrover committed:
Merge remote-tracking branch 'ftjuh/main' into fixes-1

awgrover committed:
Implement storing address-change on samd (flash, so reset on code
upload).
firmware/firmware.ino:
add eeprom emulator for samd.
more verbose about reading/writing stored address.
lengthen serial wait to allow arduino-ide to reopen serial console.
factor setup_wire() to re-use.
live address change if architectur supports it.
firmware/firmware_modules.h:
enable pini2c by default.
library.properties:
add "samd" to architectures.
add dependency on FlashStorage (eeprom emulator).
src/I2Cwrapper.h:
add getAddress()

(cherry picked from commit 82031ed9c0b986fab9624ab467bb1b297e154f37)

ftjuh committed:
improved I2C delay time taking

ftjuh committed:
added I2C state machine

ftjuh committed:
refined auto adjust incl. example

ftjuh committed:
introduced feature modules, _statusLED module, added new injection stages

ftjuh committed:
added README section for feature modules/Status LED module

ftjuh committed:
added info on state machine and state flow diagram in docs

ftjuh committed:
minor edits

ftjuh committed:
Merge branch 'feature/selfadjustI2Cdelay' into main

ftjuh committed:
updated doxygen documentation

ftjuh committed:
minor doc edits

ftjuh committed:
corrected command offset in TM1638liteI2C.h

ftjuh committed:
added ATtiny buffer limit warning in readme

ftjuh committed:
added comment for LED_BUILTIN not defined for some boards

ftjuh committed:
handle LED_BUILTIN not defined for some boards

ftjuh committed:
removed min() max() macro uses to avoid ESP32 inconsistencies

ftjuh committed:
Cleaned up and unified ESP* IRAM_ATTR stuff

ftjuh committed:
Eliminated hardware reset; added re-inizialization code for all modules and core firmware

ftjuh committed:
completeted reset/reinitialization code with I2C target restart via Wire.end()

ftjuh committed:
added optional delay parameter to I2Cwrapper::reset()

ftjuh committed:
removed delay() after reset() from all examples, it is now included in the controller's reset function

ftjuh committed:
moved default I2C address from firmware.ino to I2Cwrapper.h (I2CwrapperDefaultAddress)

ftjuh committed:
implemented I2C address feature modules; adapted Change_address demo

ftjuh committed:
minor edits; applied Arduino auto format

ftjuh committed:
Merge branch 'feature/addressconfig' into main

ftjuh committed:
added example setup image to docs

ftjuh committed:
added example setup image to docs

ftjuh committed:
minor doc edits

ftjuh committed:
README/doxygen conciliation

ftjuh committed:
docs updated

ftjuh committed:
removed diagnostics code

ftjuh committed:
removed diagnostics code

ftjuh committed:
prepare release v0.3.0

ftjuh committed:
minor doc edit

ftjuh committed:
test doxy/readme internal links

ftjuh committed:
fixed all doxygen error inducing problems

ftjuh committed:
reintroduced absolute links to repo in readme to make them work in doxygen, too

ftjuh committed:
more links corrected in readme

ftjuh committed:
more links corrected in readme

ftjuh committed:
completed template source files' documentation

ftjuh committed:
rebuilt doxy docs from scratch

ftjuh committed:
added reset documentation for modules

ftjuh committed:
readme todo list update & minor edits

ftjuh committed:
constructor maxBuf parameter comment update

ftjuh committed:
doxygen docs updated

ftjuh committed:
minor readme edits

ftjuh committed:
updated source documentation

ftjuh committed:
there is no attiny architecture, ATtinys are avr it seems

ftjuh committed:
doxy docs updated

ftjuh committed:
updated installing instructions

ftjuh committed:
minor corrections

awgrover committed:
Merge remote-tracking branch 'ftjuh/main' into fixes-1

awgrover committed:
fix #endif comments. one #include quote

awgrover committed:
samd section to readme

awgrover committed:
fix #endif

awgrover committed:
add itsy m4 to samd section

awgrover committed:
repair an ifdef section

awgrover committed:
revert an edit

awgrover committed:
log a warning for analogReference

awgrover committed:
rewrite SAMD section

awgrover committed:
remove serial wait from example, remove ifdef SerialUSB from firmware but keep serial wait

awgrover committed:
remove library dependency
